Add line number support in the sabberworm CSS parser

sidkshatriya opened the pull request:
AMP HTML is a restricted subset of HTML (See https://www.ampproject.org/ for more details) and the AMP PHP Library provide validation and HTML to AMP HTML conversion utilities. We're using the `masterminds/html5-php` library to parse HTML5 and we're using `sabberworm/php-css-parser` to parse the CSS in the [AMP PHP Library](https://github.com/Lullabot/amp-library)

After parsing the CSS parse we report any problems in the CSS (for e.g. `!important` is disallowed in the AMP dialect of HTML, we use the `sabberworm` parser to find all the uses of `!important` for instance)

Now the CSS supplied by the user could be hundreds / thousands of lines long. So its not useful to simply say that `!important` was used and is not allowed. We should be able to say _which line_ the problem occurred.

This pull request adds support for line numbers. Here is how we're using this https://github.com/Lullabot/amp-library/pull/80#issuecomment-222695630
Notice the line numbers reported in the validation errors which coming from code submitted in this PR.

Would greatly appreciate if you had a look and merge it if you believe this feature would be useful. Hopefully the code changes have not missed anything out.

Thanks for this useful library BTW.

sabberworm commented:
Wow, thanks for your work. I will review soon…

Maybe you could add a unit test dealing specifically with line numbers and refactor the one you changed to ignore the line numbers so it doesn’t break when we change the source?

sabberworm commented on lib/Sabberworm/CSS/Parser.php:
```diff
@@ -516,13 +519,18 @@ private function peek($iLength = 1, $iOffset = 0) {
 
 	private function consume($mValue = 1) {
 		if (is_string($mValue)) {
+			$noLines = substr_count($mValue, "\n");
```
Call the variable `$iLineCount` for consistency.

sidkshatriya replied:
Done

sabberworm commented on lib/Sabberworm/CSS/Parser.php:
```diff
@@ -516,13 +519,18 @@ private function peek($iLength = 1, $iOffset = 0) {
 			$this->iCurrentPosition += $this->strlen($mValue);
 			return $mValue;
 		} else {
+			$substring = $this->substr($this->iCurrentPosition, $mValue);
```
Why get the substring here when we already have it as `$sResult` two lines down?

sidkshatriya replied:
Fixed

sabberworm commented:
Add the `getLineNum` method declaration to the `Renderable` interface.
And name it either `getLineNo` or `getLineNumber`.

The `CSSList` class should also implement `Renderable`, I forgot to do that.

sabberworm commented on lib/Sabberworm/CSS/Parser.php:
```diff
@@ -516,13 +519,18 @@ private function peek($iLength = 1, $iOffset = 0) {
+			$noLines = substr_count($substring, "\n");
+			$this->iLineNum += $noLines;
 			if ($this->iCurrentPosition + $mValue > $this->iLength) {
 				throw new UnexpectedTokenException($mValue, $this->peek(5), 'count');
```
Add line numbers to exceptions as well for easier debugging.

sidkshatriya replied:
Fixed

sabberworm commented on lib/Sabberworm/CSS/CSSList/CSSList.php:
```diff
@@ -26,11 +24,19 @@ public function __construct($iLineNum = 0) {
-	public function getLineNum()
+	public function getLineNo()
 	{
 		return $this->iLineNum;
```
Rename the variable, too.

sidkshatriya replied:
Done

sabberworm commented on lib/Sabberworm/CSS/CSSList/CSSList.php:
```diff
@@ -26,11 +24,19 @@ public function __construct($iLineNum = 0) {
 	 * @return int
 	 */
-	public function getLineNum()
+	public function getLineNo()
```
Put the opening brace on the same line.

sabberworm commented:
Thanks. I do have one minor nitpick: the `Renderable` interface should not declare the setter for the line number, only the getter because in order for something to be renderable, the line-number need not be state, it could also be computed upon calling.

sabberworm commented on lib/Sabberworm/CSS/Parsing/OutputException.php:
```diff
@@ -6,4 +6,13 @@
 * Thrown if the CSS parsers attempts to print something invalid
 */
 class OutputException extends \Exception {
+    private $iLineNum;
```
Provide a getter for the line number.

sidkshatriya replied:
Done

sabberworm commented on lib/Sabberworm/CSS/Parsing/OutputException.php:
```diff
@@ -6,4 +6,13 @@
 */
 class OutputException extends \Exception {
+    private $iLineNum;
+    public function __construct($sMessage, $iLineNum = 0)
```
Put opening brace on same line as method declaration.

sidkshatriya replied:
Done

sidkshatriya commented:
Some notes:
- The `OutputException` and `UnexpectedTokenException` exception classes now have line number information (see the throws) but I think there are 4 places where you're just throwing `new \Exception` -- it would logical to add line numbers there too -- but I'm not sure if should create another custom exception class or re-use the `UnexpectedTokenException` (please let me know what params one must use)
- CSSList implements Renderable now. **In addition Rule also implements Renderable now**

> Thanks. I do have one minor nitpick: the Renderable interface should not declare the setter for the line number, only the getter because in order for something to be renderable, the line-number need not be state, it could also be computed upon calling.

Its a good idea to have setLineNo generally, mainly for testing purposes where you want to reset the stored line number to 0. What is your alternate suggestion?

sabberworm commented on lib/Sabberworm/CSS/Parsing/OutputException.php:
```diff
@@ -7,12 +7,15 @@
     }
+
+    public function getLineNo() {
+        return $this->getLine();
```
I meant for the CSS line numner (`$iLineNo` member), not for the line number of the PHP code that threw the exception.

sidkshatriya replied:
@sabberworm Good catch! I've fixed that.

Anyways, I renamed `$iLineNum` to `$iLineNo` everywhere -- is that OK? Or should I revert to the older `$iLineNum` ?

sabberworm replied:
> On 30.06.2016, at 11:37, Sidharth Kshatriya <notifications@github.com> wrote:
> 
> In lib/Sabberworm/CSS/Parsing/OutputException.php:
> 
> >          $this->$iLineNo = $iLineNo;
> >          if (!empty($iLineNo)) {
> >              $sMessage .= " [line no: $iLineNo]";
> >          }
> >          parent::__construct($sMessage);
> >      }
> > +
> > +    public function getLineNo() {
> > +        return $this->getLine();
> 
> @sabberworm Good catch! I've fixed that.
> 
> Anyways, I renamed $iLineNum to $iLineNo everywhere -- is that OK? Or should I revert to the older $iLineNum ?

No, it’s perfect.

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sabberworm commented:
Regarding the exceptions:

I used the generic `Exception` class when we encounter state we can’t possibly recover from. Maybe we should add a class `SourceException` (exception in source code) which has a line number and use it in place of `Exception`. We could then let `OutputException` and `UnexpectedTokenException` inherit from that. This should even be backwards-compatible unless someone explicitly uses `get_class` to get the type of an exception instead of using `instanceof`.

Regarding the line number setter: I would have suggested just removing it from the interface. I was fine with leaving it in the implementations. But seeing the setters are only used in the unit test, I’m thinking we should get rid of them altogether and change this

	$oColor->setLineNo(0);
	$this->assertEquals(array('r' => new Size(35.0, null, true), 'g' => new Size(35.0, null, true), 'b' => new Size(35.0, null, true)), $oColor->getColor());

into this:

	$this->assertEquals(array('r' => new Size(35.0, null, true, $oColor->getLineNo()), 'g' => new Size(35.0, null, true, $oColor->getLineNo()), 'b' => new Size(35.0, null, true, $oColor->getLineNo())), $oColor->getColor());

sidkshatriya commented:
@sabberworm 
- We pass in the line number as suggested so that the assertEqual has the correct value for the color test
- Removed `setLineNo()` from everywhere
- Added new class `SourceException`

sabberworm commented on lib/Sabberworm/CSS/Parsing/UnexpectedTokenException.php:
```diff
@@ -28,11 +28,7 @@ public function __construct($sExpected, $sFound, $sMatchType = 'literal', $iLine
+		parent::__construct($sMessage, $iLineNo);
 	}
 
 	public function getLineNo() {
```
Remove `getLineNo` and `private $iLineNo;`.

sidkshatriya replied:
@sabberworm Good catch! Done!

sabberworm commented:
Do you want to add unit tests for line numbers or shall I? I can do it next week and then release. Does that work for you?

sidkshatriya commented:
@sabberworm If I add the test myself in sometime would the timelines be earlier? When might you merge to master? When might you create a release tag? (I'm asking because I would need to change some `composer.json` dependencies accordingly for the AMP PHP Library)

No hurries, of course, but I just wanted to know your plan...

sabberworm commented:
If you don’t add unit tests, I will merge to master now and add the unit tests next week. If you intend to add unit tests now, I will wait with merging until you’re done and release immediately afterwards.

sidkshatriya commented:
@sabberworm OK -- I'll add a test now... (Thanks for being super responsive on this PR -- Appreciate it)

sabberworm commented on tests/files/line-numbers.css:
```diff
@@ -0,0 +1,28 @@
+@charset "utf-8"; /* line 3 */
```
Comment should say `1`

sidkshatriya replied:
Changed

sabberworm commented on tests/files/line-numbers.css:
```diff
@@ -0,0 +1,28 @@
+@charset "utf-8"; /* line 3 */
+
+@namespace "http://toto.example.org"; /* line 1 */
```
Comment should say `3`

sidkshatriya replied:
Changed

sabberworm commented on lib/Sabberworm/CSS/Parser.php:
```diff
@@ -232,13 +236,13 @@ private function parseStringValue() {
 				$sContent = $this->parseCharacter(false);
 				if ($sContent === null) {
-					throw new \Exception("Non-well-formed quoted string {$this->peek(3)}");
+					throw new \Exception("Non-well-formed quoted string {$this->peek(3)}", $this->iLineNo);
```
Change to `SourceException`

sabberworm commented on tests/files/line-numbers.css:
```diff
@@ -0,0 +1,28 @@
+
+body {
+    background: #FFFFFF url("http://somesite.com/images/someimage.gif") repeat top center; /* line 25 */
+    color: rgb(233, 100, 450);
```
Could you maybe split up the color components into separate lines and check if they are identified correctly:

```css
color: rgb( /* line 26 */
       233, /* line 27 */
       100, /* line 28 */
       450 /* line 29 */
);
```

sabberworm commented:
@sidkshatriya With any luck we can drop PHP 5.3 somewhere in the next year…

sabberworm commented on lib/Sabberworm/CSS/Parsing/SourceException.php:
```diff
@@ -0,0 +1,18 @@
+class SourceException extends \Exception {
+    private $iLineNo;
+    public function __construct($sMessage, $iLineNo = 0) {
+        $this->$iLineNo = $iLineNo;
```
Remove the `$`.

sidkshatriya replied:
@sabberworm Why should we remove the `$` -- sorry I didn't understand...

sabberworm replied:
> On 30.06.2016, at 14:45, Sidharth Kshatriya <notifications@github.com> wrote:
> 
> @sabberworm Why should we remove the $ -- sorry I didn't understand...
> 
> 

It’s just a typo. The line reads `$this->$iLineNo = $iLineNo;` but should be `$this->iLineNo = $iLineNo;`.

If $iLineNo is `1`, the former sets `$this->1` to `1`, the latter sets `$this->iLineNo` to `1`.

The newly-added test `testUnexpectedTokenExceptionLineNo` should have caught this but didn’t because you didn’t `use Sabberworm\CSS\Parsing\UnexpectedTokenException`.

sabberworm replied:
It’s just a typo. The line reads `$this->$iLineNo = $iLineNo;` but should be `$this->iLineNo = $iLineNo;`.

If $iLineNo is `1`, the former sets `$this->1` to `1`, the latter sets `$this->iLineNo` to `1`.

The newly-added test `testUnexpectedTokenExceptionLineNo` should have caught this but didn’t because you didn’t `use Sabberworm\CSS\Parsing\UnexpectedTokenException`.

sidkshatriya replied:
@sabberworm This was a fantastic catch! Sorry I didn't notice the extraneous `$`

sabberworm commented:
- [ ] Add some tests for exceptions, e.g.:

```php
/**
* @expectedException Sabberworm\CSS\Parsing\UnexpectedTokenException
*/
public function testUnexpectedTokenExceptionLineNo() {
	$oParser = new Parser("\ntest: 1;", Settings::create()->beStrict());
	try {
		$oParser->parse();
	} catch(UnexpectedTokenException $e) {
		$this->assertSame($e->getLineNo(), 2);
		throw $e;
	}
}
```

Ideally there would be a test for every place in the parser where we throw an exception.

sidkshatriya commented:
@sabberworm Hopefully things are now looking good. Let me know if you have any other suggestions (thanks for your really responsive help on this!). 

BTW My contributions on this pull request are on behalf of https://github.com/Lullabot

I'll pick up with this tomorrow (or later tonight if I get the chance)

sabberworm commented on tests/Sabberworm/CSS/ParserTest.php:
```diff
@@ -435,4 +438,70 @@ function parsedStructureForFile($sFileName, $oSettings = null) {
+		try {
+			$oParser->parse();
+		} catch (UnexpectedTokenException $e) {
+			$this->assertSame($e->getLineNo(), 2);
```
Switch the arguments: expected should come first, actual later.

sidkshatriya replied:
@sabberworm done.

sabberworm commented on tests/Sabberworm/CSS/ParserTest.php:
```diff
@@ -435,4 +438,70 @@ function parsedStructureForFile($sFileName, $oSettings = null) {
+	/**
+	 * @expectedException Sabberworm\CSS\Parsing\UnexpectedTokenException
+	 */
+	public function testUnexpectedTokenExceptionLineNo() {
```
You can remove `public`.

sidkshatriya replied:
Done

sabberworm commented on tests/Sabberworm/CSS/ParserTest.php:
```diff
@@ -435,4 +438,70 @@ function parsedStructureForFile($sFileName, $oSettings = null) {
+		$oParser = new Parser("\ntest: 1;", Settings::create()->beStrict());
+		try {
+			$oParser->parse();
+		} catch (UnexpectedTokenException $e) {
```
Either use the full class namespace `Sabberworm\CSS\Parsing\UnexpectedTokenException` or `use Sabberworm\CSS\Parsing\UnexpectedTokenException` at the top of the file.

sidkshatriya commented:
@sabberworm Ah you merged! Great! Thanks so much!

sabberworm commented:
> On 30.06.2016, at 16:41, Sidharth Kshatriya <notifications@github.com> wrote:
> 
> @sabberworm Ah you merged! Great! Thanks so much!

Thank you for your great work! Now I only need someone like you to implement selector and media query parsing…

How do you prefer to be listed under contributors, as individual or as Lullabot?

sidkshatriya commented:
as Lullabot

**re:**  selector and media query parsing -- I could look at it in my spare time. If you make a ticket chalking out the details on whats incomplete / what needs to be done etc. it would be awesome. What happens with media queries right now? Does the parser choke or simply not get the finer structure?

sabberworm commented:
The selector parsing is really rudimentary, it just searches for the next `{` and splits by comma. So it has trouble parsing a selector `like a[href*=","] .complicated\{selector`.

Media query parsing is even worse. It was only designed to handle `@media print {}` and chokes on more complex stuff. See #75 and its duplicates (#85, #104) for examples.

sabberworm commented:
Fixing media queries should be a lot simpler than fixing selectors because there are fewer cases to consider and since the current selector parsing works well enough for real-world, non-contrived examples, it’s also the bigger pain point.

sidkshatriya commented:
@sabberworm 
Is your parser hand built or have you consulted something like https://drafts.csswg.org/css-syntax-3/ for your tokenizing/parsing algorithm?

I don't really need media query parsing or selector parsing for the AMP library at the moment (may need it in future) -- but I'm concerned if the parser will abort and issue an exception. Its OK for instance to not do an accurate job of parsing for things like selectors/media queries but aborting the whole parse is not so good... What happens in such cases?

sidkshatriya commented:
@sabberworm sorry for persisting but I just want to be aware of any parse related aborts due to syntax the parser does not understand. Like I mentioned, it's ok if you don't parse the complex selectors into its fine components or even get media query parsing correct. I just want the parser to skip over those things (or even do a bad job with them). What _is_ problematic is the whole parse getting aborted due to unrecognised (but valid) css syntax.

sabberworm commented:
Some well-defined exception states are recoverable. But because the parser does not know how media queries are supposed to be structures, it can’t recover and throws a `SourceException`. But I will try to implement correct media query parsing soon since it’s long been one of the major shortcomings of this library.

sidkshatriya commented:
@sabberworm Thanks for responding and eagerly look forward to some basic media query parsing. 

Also to my previous question: is your parsing algorithm hand built or you have consulted something like https://drafts.csswg.org/css-syntax-3/ ? IIRC that there may be general strategies in that link where you can prevent yourself from aborting a parse when you encounter syntax that you don't know about. Basically you should simply skip over to the next at-rule or declaration block that you _do_ understand.

sabberworm commented:
The goal of the parser was always to parse valid CSS. Handling invalid CSS and being fault-tolerant was a feature added later on and does not exactly follow the way the specification demands of browser vendors.

However, for valid CSS (for everything the parser supports, which, currently, does not include media queries) the goal was always to be compliant with the specs, even if the resulting data structure does not adhere to CSSOM.
